ide/posts/2020-06-50-ghcide-and-nixpkgs.md

```diff
@@ -0,0 +1,321 @@
+   a `ghc` 8.8 project.
+2. Using the same `ghc` release but with a patch to `ghc`. This happened to me
+   while using the `obelisk` framework which uses a modified `ghc`.
+3. Using the same `ghc` version but linked against different external libraries
+   like `glibc`. This can happen when different releases of nixpkgs are
+   involved.
+
+To sum up, both ghcs should come from the same source and be linked against the
+same libraries. Your best bet is to use the same binary. But that is not
+necessary.
+
+### How to actually get a ghcide on nix
+
+So consider the following situation. You are using nix to develop your Haskell
+project. To be more specific: To compile your project you use ghc installed
+from nixpkgs. This probably means that you have a `default.nix` and a
+`shell.nix` file in your project.
+
+So what are your ways to get a ghcide?
+
+1. Use the `ghcide-nix` derivation.
+2. Install `haskellPackages.ghcide` from nixpkgs.
+3. Compile one yourself using cabal.
+
+So I was developing an obelisk project and wanted to use ghcide. The first way
+was the recommend one in the `ghcide` README, but sadly that didn‘t work
+because (as I learned later from Elliot Cameron) the obelisk project uses a
+patched ghc. No luck here. We actually tried to pass the patched `ghc` into the
+`ghcide-nix` derivation, but `haskell.nix` refuses to work with a not
+`haskell.nix` ghc. So this was getting complicated quickly.
+2. Was also not an option, since `haskellPackages.ghcide` was marked broken in
+   nixpkgs and ignoring the flag just gave me tons of version mismatch errors.
+   So that left option three: Compiling with cabal myself. That worked-ish,
+   showing me that it‘s at least possible to get ghcide to run, but it is a
+   very unclean and brittle solution.
+
+After deciding that one and three can‘t help me I resolved to figure out how to
+fix the second approach.
+
+## The nixpkgs Haskell ecosytem and dependency resolution
+
+So I was wondering: Why can cabal get the dependencies right and nixpkgs can‘t?
+So for those of you who don‘t know: Let me digress on a short exploration of
+the nixpkgs Haskell ecosystem.
+
+### Haskell dependency resolution in general
+
+Dependency resolution problems have a long history in Haskell. But today there
+are two solutions that both work quite well in general.
+
+1. Specify upper and lower bounds for every dependency in your cabal file and
+   let cabal figure out a build plan. The times of cabal hell are over and this
+   works quite well. Notably this is the way ghcide is supposed to be compiled
+   in general.
+2. Pin a stack LTS release for your dependencies and pin the version for
+   packages not on stackage.
+
+Now solution two is in some sense less complex to use, because at compile time
+you don‘t need to construct a build plan. Of course, as I said, today cabal can
+do this for you very smoothly, which is why I personally prefer the first
+   approach.
+
+### Haskell in nixpkgs - pkgs.hackagePackages
```

Comment:
The post here is getting a bit far away from the original goal. It would be good to add some signposts which link it back to the issue with ghcide.

Reply:
I swapped it with the later part and tried to make the structure much clearer.